### PR TITLE
Alert rule project backfill migration

### PR DESCRIPTION
**Pull request by nhsiehgit**
Backfill migration to create `AlertRuleProject`s for all AlertRules

Update `create_alert_rule` method to create AlertRuleProject for all new alert rules

Add TODO's for removing projects from query subscription

migration:
Iterate through all QuerySubscriptions and create a new AlertRuleProject given the subscription.projects

### CODE REVIEW

File: src/sentry/migrations/0682_alert_rule_project_backfill_migration.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        alert_rule = snuba_query.sentry_alertrule
+
+        AlertRuleProjects.objects.create(
+            alert_rule=alert_rule,
+            project=subscription.project,
+        )
+
+
+class Migration(CheckedMigration):
+    is_dangerous = False
```
Comment (wedamija) [inline src/sentry/migrations/0682_alert_rule_project_backfill_migration.py:35]: This should be dangerous so we can run it post deploy

File: src/sentry/migrations/0682_alert_rule_project_backfill_migration.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        snuba_query = subscription.snuba_query
+        if not snuba_query:
+            logger.warning(
+                "QuerySubscription found with no snuba_query",
+                extra={"query_subscription_id": subscription.id},
+            )
+            continue
+
+        alert_rule = snuba_query.sentry_alertrule
```
Comment (wedamija) [inline src/sentry/migrations/0682_alert_rule_project_backfill_migration.py:26]: Does this work? I'm not sure this name is correct

**Review by wedamija — COMMENTED** [review]
We should have a migration test for this.

I think these logic changes in the codebase should be split out from this pr

File: src/sentry/incidents/logic.py [hunk 3]
```diff
@@ -590,19 +590,17 @@ def create_alert_rule(
...
-            ]
-            AlertRuleProjects.objects.bulk_create(arps)
 
         if monitor_type == AlertRuleMonitorType.ACTIVATED and activation_condition:
             # NOTE: if monitor_type is activated, activation_condition is required
             AlertRuleActivationCondition.objects.create(
                 alert_rule=alert_rule, condition_type=activation_condition.value
             )
 
+        # initialize projects join table for alert rule
```
Comment (nhsiehgit) [inline src/sentry/incidents/logic.py:600]: TODO: split this into a separate PR

Comment (saponifi3d) [reply src/sentry/incidents/logic.py:603]: still want to pull t his into another pr?

Comment (nhsiehgit) [reply src/sentry/incidents/logic.py:603]: pulled into https://github.com/getsentry/sentry/pull/68022

File: src/sentry/incidents/models/alert_rule.py [hunk 4]
```diff
@@ -355,6 +355,9 @@ def subscribe_projects(
         # On activated subscription, additional query parameters will be added to the constructed query in Snuba
         created_subscriptions = []
         if self.monitor_type == monitor_type.value:
+            # TODO: remove projects from QuerySubscriptions
+            # This means we will no longer need to utilize `bulk_create_snuba_subscriptions` as there's no value
+            # in iterating over projects to call `create_snuba_subscription` per project
```
Comment (wedamija) [inline src/sentry/incidents/models/alert_rule.py:360]: Hmm, why are we removing projects from `QuerySubscriptions`?

Remember that the subscription system doesn't know the other systems exist and is generic. If it doesn't know which subscription a project belongs to, it doesn't have enough information to recreate a subscription.

File: src/sentry/migrations/0685_alert_rule_project_backfill_migration.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+from sentry.new_migrations.migrations import CheckedMigration
+
+logger = logging.getLogger(__name__)
+
+
+def _backfill_alert_rule_projects(apps, schema_editor):
+    QuerySubscriptions = apps.get_model("sentry", "QuerySubscription")
+    AlertRuleProjects = apps.get_model("sentry", "AlertRuleProjects")
+
+    for subscription in QuerySubscriptions.objects.all():
```
Comment (wedamija) [inline src/sentry/migrations/0685_alert_rule_project_backfill_migration.py:16]: Probably good to use `RangeQuerySetWrapper` here

File: src/sentry/migrations/0685_alert_rule_project_backfill_migration.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                extra={"query_subscription_id": subscription.id},
+            )
+            continue
+
+        alert_rule = snuba_query.alertrule_set.get()
+
+        existing_alert_rule_projects = AlertRuleProjects.objects.filter(alert_rule=alert_rule)
+        should_create_new = not existing_alert_rule_projects.exists()
+
+        if existing_alert_rule_projects.exists() and (
```
Comment (wedamija) [inline src/sentry/migrations/0685_alert_rule_project_backfill_migration.py:31]: `not should_create_new` instead of making the `.exists()` again.

File: src/sentry/migrations/0685_alert_rule_project_backfill_migration.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            len(existing_alert_rule_projects) > 1
```
Comment (wedamija) [inline src/sentry/migrations/0685_alert_rule_project_backfill_migration.py:32]: The exists already proves this, we can remove this line

File: src/sentry/migrations/0685_alert_rule_project_backfill_migration.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            or existing_alert_rule_projects.get().project != subscription.project
```
Comment (wedamija) [inline src/sentry/migrations/0685_alert_rule_project_backfill_migration.py:33]: Is there guaranteed to be 0-1 rows here? I thought there could be many?

File: src/sentry/migrations/0685_alert_rule_project_backfill_migration.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        if not snuba_query:
+            logger.warning(
+                "QuerySubscription found with no snuba_query",
+                extra={"query_subscription_id": subscription.id},
+            )
+            continue
+
+        alert_rule = snuba_query.alertrule_set.get()
+
+        existing_alert_rule_projects = AlertRuleProjects.objects.filter(alert_rule=alert_rule)
```
Comment (wedamija) [inline src/sentry/migrations/0685_alert_rule_project_backfill_migration.py:28]: It might make sense to just coerce this to a list instead and operate on the model instances, since you're performing a lot of different operations here

File: src/sentry/incidents/models/alert_rule.py (hunk 4 above)
Comment (nhsiehgit) [reply src/sentry/incidents/models/alert_rule.py:360]: 🤔 fair point.
Yea - i guess we could either pass `project` wherever we're "create_subscription_in_snuba"ing (whichcurrently pulls projects out of the subscription instance)
OR - we maintain duplicate references to the project in QuerySubscription and AlertRule - referencing the subscription project _only_ for snuba subscription creation, while referencing AlertRule project for everything else

Ok - i can remove this comment with a caution note on the duplication 👀

File: src/sentry/migrations/0685_alert_rule_project_backfill_migration.py (hunk 7 above)
Comment (nhsiehgit) [reply src/sentry/migrations/0685_alert_rule_project_backfill_migration.py:32]: Not exactly - we're checking here that there is more than just 1 project fk'd

if there's only 1 project, then that's fine

File: src/sentry/migrations/0685_alert_rule_project_backfill_migration.py (hunk 8 above)
Comment (nhsiehgit) [reply src/sentry/migrations/0685_alert_rule_project_backfill_migration.py:33]: Yea - right now its set up to have 1 to many

but for now - there can only be a single project and we don't have any logic set for 1:many

File: src/sentry/migrations/0686_alert_rule_project_backfill_migration.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        ):
+            logger.warning(
+                "AlertRuleProject found with different project than subscription",
+                extra={
+                    "alert_rule_id": alert_rule.id,
+                    "project_count": len(existing_alert_rule_projects),
+                },
+            )
+            for arp in existing_alert_rule_projects:
+                arp.delete()
```
Comment (wedamija) [inline src/sentry/migrations/0686_alert_rule_project_backfill_migration.py:45]: Do you really want to delete all the rows here, or just the invalid ones?

File: src/sentry/migrations/0686_alert_rule_project_backfill_migration.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+            continue
+
+        alert_rule = snuba_query.alertrule_set.get()
+
+        existing_alert_rule_projects = list(AlertRuleProjects.objects.filter(alert_rule=alert_rule))
+        should_create_new = not existing_alert_rule_projects.exists()
+
+        if not should_create_new and (  # eg. if alert rule already has fk projects
+            len(existing_alert_rule_projects) > 1
+            or existing_alert_rule_projects[0].project != subscription.project
```
Comment (wedamija) [inline src/sentry/migrations/0686_alert_rule_project_backfill_migration.py:35]: `existing_alert_rule_projects[0].project_id` save you a query.

Shouldn't we be checking all rows to see if the project doesn't match?

File: src/sentry/migrations/0686_alert_rule_project_backfill_migration.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+            logger.warning(
+                "QuerySubscription found with no snuba_query",
+                extra={"query_subscription_id": subscription.id},
+            )
+            continue
+
+        alert_rule = snuba_query.alertrule_set.get()
+
+        existing_alert_rule_projects = list(AlertRuleProjects.objects.filter(alert_rule=alert_rule))
+        should_create_new = not existing_alert_rule_projects.exists()
```
Comment (wedamija) [inline src/sentry/migrations/0686_alert_rule_project_backfill_migration.py:31]: This can just be `bool(existing_alert_rule_projects)`

File: src/sentry/migrations/0686_alert_rule_project_backfill_migration.py (hunk 11 above)
Comment (nhsiehgit) [reply src/sentry/migrations/0686_alert_rule_project_backfill_migration.py:35]: the AlertRuleProject doesn't have a `project_id` attr though?
or do you mean i should be comparing the `project_id`s as opposed to the project?

I guess that's one way to do it
we could iterate through all the projects if the len(projects) >=1
Then delete any projects that don't match the subscription

now that we've listified the query, i suppose that makes sense 👍 
will update

**Review by wedamija — APPROVED** [review]
